pkg/restore/log_client.go

```diff
@@ -0,0 +1,1070 @@
+	for tableID, puller := range l.eventPullers {
+		pullerReplica := puller
+		tableIDReplica := tableID
+		workerPool.ApplyOnErrorGroup(eg, func() error {
+			return l.restoreTableFromPuller(ectx, tableIDReplica, pullerReplica, dom)
+		})
+	}
+	return eg.Wait()
+}
+
+// RestoreLogData restore specify log data from storage.
+func (l *LogClient) RestoreLogData(ctx context.Context, dom *domain.Domain) error {
+	// 1. Retrieve log data from storage
+	// 2. Find proper data by TS range
+	// 3. Encode and ingest data to tikv
+
+	// parse meta file
+	data, err := l.restoreClient.storage.Read(ctx, metaFile)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = json.Unmarshal(data, l.meta)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	log.Info("get meta from storage", zap.Binary("data", data))
+
+	if l.startTS > l.meta.GlobalResolvedTS {
+		return errors.Errorf("start ts:%d is greater than resolved ts:%d",
+			l.startTS, l.meta.GlobalResolvedTS)
+	}
+	if l.endTs > l.meta.GlobalResolvedTS {
+		log.Info("end ts is greater than resolved ts,"+
+			" to keep consistency we only recover data until resolved ts",
+			zap.Uint64("end ts", l.endTs),
+			zap.Uint64("resolved ts", l.meta.GlobalResolvedTS))
+		l.endTs = l.meta.GlobalResolvedTS
+	}
+
+	// collect ddl files
+	ddlFiles, err := l.collectDDLFiles(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	log.Info("collect ddl files", zap.Any("files", ddlFiles))
+
+	err = l.doDBDDLJob(ctx, ddlFiles)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	log.Debug("db level ddl executed")
+
+	// collect row change files
+	rowChangesFiles, err := l.collectRowChangeFiles(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	log.Info("collect row changed files", zap.Any("files", rowChangesFiles))
+
+	// create event puller to apply changes concurrently
+	for tableID, files := range rowChangesFiles {
+		name := l.meta.Names[tableID]
+		schema, table := parseQuoteName(name)
+		log.Info("create puller for table",
+			zap.Int64("table id", tableID),
+			zap.String("schema", schema),
+			zap.String("table", table),
+		)
+		l.eventPullers[tableID], err = cdclog.NewEventPuller(ctx, schema, table, ddlFiles, files, l.restoreClient.storage)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		// use table name to get table info
+		var tableInfo titable.Table
+		var allocs autoid.Allocators
+		infoSchema := dom.InfoSchema()
+		if infoSchema.TableExists(model.NewCIStr(schema), model.NewCIStr(table)) {
+			tableInfo, err = infoSchema.TableByName(model.NewCIStr(schema), model.NewCIStr(table))
+			if err != nil {
+				return errors.Trace(err)
+			}
+			dbInfo, ok := dom.InfoSchema().SchemaByName(model.NewCIStr(schema))
+			if !ok {
+				return errors.Errorf("schema %s not exists", schema)
+			}
+			allocs = autoid.NewAllocatorsFromTblInfo(dom.Store(), dbInfo.ID, tableInfo.Meta())
+		}
+
+		l.tableBuffers[tableID] = cdclog.NewTableBuffer(tableInfo, allocs, l.concurrencyCfg.BatchFlushKVPairs)
```

Comment:
If the `infoSchema.TableExists` is false, then the tableInfo should be nil here, should add a check here

Reply:
yes, the check already in `NewTableBuffer`

Reply:
Then why still need the `infoSchema.TableExists` check

Reply:
If `infoSchema.TableExists` we can build allocators and tableInfo at the beginning.
if not exists, then we can not handle allocators and tableInfo here, we should wait for next DDL happen to Reload.